src/gui/component.rs

```diff
@@ -214,39 +240,42 @@ impl Component for Greeter {
                 },
                 #[template_child]
                 user_toggle {
-                    #[track(model.updates.changed(Updates::password_mode()))]
-                    set_sensitive: !model.updates.password_mode,
+                    #[track(model.updates.changed(Updates::input_mode()))]
+                    set_sensitive: !model.updates.is_input(),
                     connect_clicked => Self::Input::ToggleManualUser,
                 },
                 #[template_child]
                 sess_toggle {
-                    #[track(model.updates.changed(Updates::password_mode()))]
-                    set_visible: !model.updates.password_mode,
+                    #[track(model.updates.changed(Updates::input_mode()))]
+                    set_visible: !model.updates.is_input(),
                     connect_clicked => Self::Input::ToggleManualSess,
                 },
                 #[template_child]
                 cancel_button {
-                    #[track(model.updates.changed(Updates::password_mode()))]
-                    set_visible: model.updates.password_mode,
+                    #[track(model.updates.changed(Updates::input_mode()))]
+                    set_visible: model.updates.is_input(),
                     connect_clicked => Self::Input::Cancel,
                 },
                 #[template_child]
                 login_button {
                     #[track(
-                        model.updates.changed(Updates::password_mode())
-                        && !model.updates.password_mode
+                        model.updates.changed(Updates::input_mode())
+                        && !model.updates.is_input()
                     )]
                     grab_focus: (),
                     connect_clicked[
                         sender,
-                        password_entry,
+                        secret_entry,
                         usernames_box,
                         username_entry,
                         sessions_box,
                         session_entry,
                     ] => move |_| {
                         sender.input(Self::Input::Login {
-                            password: password_entry.text().to_string(),
+                            // Here, it is sufficient to pass the text from either secret_entry or visible_entry,
+                            // because they never show at the same time and their text is synchronised through the
+                            // model.updates.input field.
+                            input: secret_entry.text().to_string(),
```

Comment:
I'm not so sure about this. Have you tested it?

IMHO `model.updates.input` should change the text of both `secret_entry` and `visible_entry`, but not the other way around (as far as my understanding of Relm4 goes). If my hunch is correct, then you'll have to use a conditional on `model.updates.input_mode()` to use either `secret_entry` or `visible_entry`.

Reply:
It does work - I tested it by changing `secret_entry` to `visible_entry` in this piece of code and using password login.

I agree that this is hacky and I actually wanted to implement your idea first - branching based on `input_mode`. However I wasn't sure how to do that given that this is in a closure that moves the variables. If you know how to access the model in the closure I'll definitely update it, otherwise I could maybe ask somewhere else for advice.

Reply:
I found out how to do this. I simply made the `InputMode` enum `Copy`, so that it can be copied inside the closure automatically.